src/Symfony/Component/Encryption/SymfonyEncryptionToken.php

```diff
@@ -0,0 +1,165 @@
+ * This class is responsible for creating and parsing Symfony Encryption Tokens.
+ * It is an helper class for the EncryptionInterface implementations. In all common
+ * scenarios, one does not need to interact with this class directly.
+ *
+ * @author Tobias Nyholm <tobias.nyholm@gmail.com>
+ */
+class SymfonyEncryptionToken implements \Stringable
+{
+    /**
+     * Algorithm used to encrypt the message.
+     */
+    private string $algorithm;
+    private string $version;
+    private string $ciphertext;
+
+    /**
+     * @var array<string, string>
+     */
+    private array $headers = [];
+
+    private function __construct()
+    {
+    }
+
+    /**
+     * @param array<string, string> $headers keys MUST be ascii values only
+     */
+    public static function create(string $algorithm, string $ciphertext, array $headers = []): self
+    {
+        $model = new self();
+        $model->algorithm = $algorithm;
+        $model->ciphertext = $ciphertext;
+        $model->headers = $headers;
+
+        return $model;
+    }
+
+    /**
+     * Take a string representation of the ciphertext and parse it into an object.
+     *
+     * @throws MalformedCipherException
+     */
+    public static function parse(string $input): self
+    {
+        $parts = explode('.', $input);
+        if (!\is_array($parts) || 3 !== \count($parts)) {
+            throw new MalformedCipherException();
+        }
+
+        [$headersString, $ciphertext, $hashSignature] = $parts;
+
+        $headersString = self::base64UrlDecode($headersString);
+        $ciphertext = self::base64UrlDecode($ciphertext);
+        $hashSignature = self::base64UrlDecode($hashSignature);
+
+        // Check if data has been modified
+        $hash = hash('sha256', $headersString.$ciphertext);
+        if (!hash_equals($hash, $hashSignature)) {
+            throw new MalformedCipherException();
+        }
+
+        $headers = json_decode($headersString, true);
+        if (!\is_array($headers) || !\array_key_exists('alg', $headers) || !\array_key_exists('ver', $headers)) {
+            throw new MalformedCipherException();
+        }
+
+        foreach ($headers as $name => $value) {
+            $headers[$name] = self::base64UrlDecode($value);
+        }
+
+        $model = new self();
+        $model->algorithm = $headers['alg'];
+        unset($headers['alg']);
+        $model->version = $headers['ver'];
+        unset($headers['ver']);
+        $model->headers = $headers;
+        $model->ciphertext = $ciphertext;
+
+        return $model;
+    }
+
+    public function __toString(): string
+    {
+        return $this->getString();
+    }
+
+    public function getString(): string
+    {
+        $headers = $this->headers;
+        $headers['alg'] = $this->algorithm;
+        $headers['ver'] = (isset($headers['ver']) && '' !== $headers['ver']) ? $headers['ver'] : '1';
+        foreach ($headers as $name => $value) {
+            $headers[$name] = self::base64UrlEncode($value);
```

Comment:
`headers` is json_encoded + bade64encoded, no need to base64 encode the value twice IMHO

Reply:
Actually, we do. 

`json_encode` does not support non-ascii values. We do `base64UrlEncode()` on each value to make sure we only have ascii values so we are able to run json_encode. 

It is common to have nonce with non-ascii values.